src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,170 @@
+                // Notifications are emitted immediately; log events buffer until the next `interval`.
+                Some(tap) = tap_rx.next() => {
+                    let tap = tap.into();
+
+                    // Emit notifications immediately; these don't count as a 'batch'.
+                    if let LogEventPayload::Notification(_) = tap {
+                        // If an error occurs when sending, the subscription has likely gone
+                        // away. Break the loop to terminate the thread.
+                        if log_tx.send(vec![tap]).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        // A simple implementation of "Algorithm R" per
+                        // https://en.wikipedia.org/wiki/Reservoir_sampling. As we're unable to
+                        // pluck the nth result, this is chosen over the more optimal "Algorithm L"
+                        // since discarding results isn't an option.
+                        if limit > results.len() {
+                            results.push(tap);
+                        } else {
+                            let random_number = rng.gen_range(0..batch);
+                            if random_number < results.len() {
+                                results[random_number] = tap;
+                            }
+                        }
+                        // Increment the batch count, to be used for the next Algo R loop
+                        batch += 1;
+                    }
+                }
+                _ = interval.tick() => {
+                    // If there are any existing results after the interval tick, emit.
+                    if !results.is_empty() {
+                        // Reset the batch count, to adjust sampling probability for the next round.
+                        batch = 0;
+
+                        // Since events will appear out of order per the random sampling
+                        // strategy, drain the existing results and sort by timestamp.
+                        let results = results
+                            .drain(..)
+                            .sorted_by(|a, b| match (a, b) {
+                                (LogEventPayload::LogEvent(a), LogEventPayload::LogEvent(b)) => {
+                                    match (a.get_timestamp(), b.get_timestamp()) {
+                                        (Some(a), Some(b)) => a.cmp(b),
+                                        _ => Ordering::Equal,
+                                    }
+                                }
+                                _ => Ordering::Equal,
+                            })
+                            .collect();
+
+                        // If we get an error here, it likely means that the subscription has
+                        // gone has away. This is a valid/common situation.
+                        if log_tx.send(results).await.is_err() {
```

Comment:
Same here, could we log this?

Reply:
Logged in 19aa7cd7e1bd9ee4a91bd8140a5e1448e9f8a7b1.